feat: extract CodegenEngine

alex-chew opened the pull request:
*Issue #, if available:* #151 (partial)

*Description of changes:* Extracts the codegen management logic from `CodegenCli` into a `CodegenEngine`, so that both `CodegenCli` and the `smithy-dafny-codegen` plugin serve as thin entrypoints to `CodegenEngine`. The `CodegenCli` interface is not expected to change.

TODO:

- [x] Fix SQS TestModel Gradle configuration so it can build
- [x] Generate csproj file
- ~~Generate Makefile~~ (will add as fast-follow)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

alex-chew committed:
extract CodegenEngine

alex-chew committed:
fix Gradle build, and some cleanup

alex-chew committed:
fix includeDafnyFile handling

alex-chew committed:
fix missing discoverModels

alex-chew committed:
fix "conflicting shape definition" for SQS model

alex-chew committed:
typo

alex-chew committed:
Merge branch 'main-1.x' into extract-codegen-engine

alex-chew committed:
generate csproj (partially)

alex-chew committed:
substitute STDLIB_PATH in csproj template

robin-aws committed:
Patch to only use the build root if includeDafnyFile is relative

robin-aws committed:
Merge branch 'extract-codegen-engine' of github.com:awslabs/polymorph into extract-codegen-engine

alex-chew committed:
cleanup

alex-chew committed:
add CI test for building SQS model via plugin

alex-chew committed:
try to fix SQS build

alex-chew committed:
try again

alex-chew committed:
typo

robin-aws reviewed (changes requested):
I still have to have a closer look at the core CodegenEngine, but otherwise this looks great aside from these comments.

robin-aws committed:
Try test model with gradle-build-action

alex-chew committed:
Dafny as explicit target lang is an error

Co-authored-by: Robin Salkeld <salkeldr@amazon.com>

alex-chew committed:
typo

Co-authored-by: Robin Salkeld <salkeldr@amazon.com>

texastony reviewed (changes requested):
Publishing an incomplete review.
Will review from a scratch when this PR is moved out of draft.

alex-chew committed:
extract plugin settings, add edition property

alex-chew committed:
Merge remote-tracking branch 'origin/extract-codegen-engine' into extract-codegen-engine

alex-chew committed:
cleanup

alex-chew committed:
add warning that Java support is experimental

alex-chew committed:
add TODO about service closure vs namespace

alex-chew committed:
try fixing CI

`gradle build` in the SQS test model directory, works on macOS but not
on Linux (in CI). hypothesis: this is because the Smithy build tooling
automatically searches for a "model" directory adjacent to the
smithy-build.json, which passes on macOS only due to the
case-insensitive-by-default filesystem.

this commit tries to work around that by explicitly specifying the
"Model" directory as a model source directory.

see: https://smithy.io/2.0/guides/building-models/build-config.html#using-smithy-build-json

alex-chew committed:
Merge branch 'main-1.x' into extract-codegen-engine

alex-chew committed:
update SQS model readme

robin-aws reviewed (commented):
LGTM aside from a couple of small nits

alex-chew committed:
Dafny as explicit target lang is an error, redux

Co-authored-by: Robin Salkeld <salkeldr@amazon.com>

alex-chew committed:
add TODO for registering manifest files

alex-chew committed:
Merge remote-tracking branch 'origin/extract-codegen-engine' into extract-codegen-engine

robin-aws committed:
Rename Model to model instead of adding Model in smithy-build.json

The latter approach doesn’t work on a case insensitive file system (at least on my macbook it led to loading models twice), but using “model” will hopefully work with the CLI as well (will push more fixes if necessary)

alex-chew committed:
make a copy of the SQS test model for CLI tests

alex-chew committed:
Merge branch 'main-1.x' into extract-codegen-engine

alex-chew committed:
trying upgrade SQS model gradlew to 7.6

alex-chew committed:
Merge remote-tracking branch 'origin/main-1.x' into extract-codegen-engine

alex-chew committed:
log more output to debug CI failure

alex-chew committed:
consolidate smithyVersion to 1.28.1

I can now reproduce the CI error locally...

alex-chew committed:
pin Smithy CLI version in SQS Gradle buildscript

alex-chew committed:
accept --java-aws-sdk-version sans --aws-sdk-style

alex-chew committed:
Merge branch 'main-1.x' into extract-codegen-engine
